File: apps/grokking-ml-python-lib/src/grokking_ml_lib/nb09_app_recommendations.py
"""原著ノートブック #09 `Chapter_09_Decision_Trees/App_recommendations.ipynb`。

6 人のユーザーの「使っている端末」と「年齢」から、おすすめアプリを決定木で当てる。
原著は同じデータを 2 通りの形で学習させ、木の形がどう変わるかを見せている。

1. 年齢を **カテゴリ**（若者 / 大人）に潰して one-hot 符号化する
2. 年齢を **数値** のまま渡す

`DecisionTreeClassifier` は同点の分割候補から **無作為に 1 つ選ぶ** ので、
木の形は実行のたびに変わる。何が安定していて何が変わるのかを分けて扱う。
"""

from __future__ import annotations

from dataclasses import dataclass

import pandas as pd
from sklearn.tree import DecisionTreeClassifier

#: 原著の元データ。この 6 人ぶんの情報しかない
APP_DATASET = pd.DataFrame(
    {
        "Platform": ["iPhone", "iPhone", "Android", "iPhone", "Android", "Android"],
        "Age": [15, 25, 32, 35, 12, 14],
        "App": [
            "Atom Count",
            "Check Mate Mate",
            "Beehive Finder",
            "Check Mate Mate",
            "Atom Count",
            "Atom Count",
        ],
    }
)

#: 年齢をカテゴリに潰して one-hot 符号化したもの。原著が手で書き下している
APP_DATASET_ONE_HOT = pd.DataFrame(
    {
        "Platform_iPhone": [1, 1, 0, 1, 0, 0],
        "Platform_Android": [0, 0, 1, 0, 1, 1],
        "Age_Young": [1, 0, 0, 0, 1, 1],
        "Age_Adult": [0, 1, 1, 1, 0, 0],
        "App_Atom_Count": [1, 0, 0, 0, 1, 1],
        "App_Beehive_Finder": [0, 0, 1, 0, 0, 0],
        "App_Check_Mate_Mate": [0, 1, 0, 1, 0, 0],
    }
)

#: カテゴリ版で使う特徴量の列。原著の順序をそのまま保つ
CATEGORICAL_FEATURES = ["Platform_iPhone", "Platform_Android", "Age_Adult", "Age_Young"]

#: カテゴリ版の目的変数。3 列を同時に予測する多出力分類になる
CATEGORICAL_LABELS = ["App_Atom_Count", "App_Beehive_Finder", "App_Check_Mate_Mate"]

#: 数値版で使う特徴量の列
NUMERIC_FEATURES = ["Age", "Platform_iPhone", "Platform_Android"]


@dataclass
class TreeSummary:
    """学習した木の形を、比較しやすい形にまとめたもの。"""

    #: 根で使われた特徴量の名前
    root_feature: str
    #: 根の分割しきい値
    root_threshold: float
    #: 根のジニ不純度
    root_gini: float
    #: 節と葉を合わせた数
    node_count: int
    #: 訓練データに対する正解率
    score: float


def _numeric_dataset() -> pd.DataFrame:
    """数値版のデータ。端末だけ one-hot にし、年齢は数値のまま残す。"""
    return pd.DataFrame(
        {
            "Platform_iPhone": [1, 1, 0, 1, 0, 0],
            "Platform_Android": [0, 0, 1, 0, 1, 1],
            "Age": APP_DATASET["Age"],
            "App": APP_DATASET["App"],
        }
    )


def fit_categorical(random_state: int | None = None) -> DecisionTreeClassifier:
    """年齢をカテゴリに潰した版を学習する。

    目的変数が 3 列あるので **多出力分類** になる。scikit-learn の決定木は
    複数の列を同時に予測でき、各節のジニ不純度は列ごとの平均になる。
    """
    features = APP_DATASET_ONE_HOT[CATEGORICAL_FEATURES]
    labels = APP_DATASET_ONE_HOT[CATEGORICAL_LABELS]
    return DecisionTreeClassifier(random_state=random_state).fit(features, labels)


def fit_numeric(random_state: int | None = None) -> DecisionTreeClassifier:
    """年齢を数値のまま渡した版を学習する。

    決定木がしきい値を自分で決めるので、`Age <= 20.0` のような分割が現れる。
    どこで切るかは、隣り合う値の中点から選ばれる（12, 14, 15 と 25 の間なら 20）。
    """
    dataset = _numeric_dataset()
    return DecisionTreeClassifier(random_state=random_state).fit(
        dataset[NUMERIC_FEATURES], dataset["App"]
    )


def summarize(model: DecisionTreeClassifier, feature_names: list[str]) -> TreeSummary:
    """木の形を [TreeSummary] にまとめる。"""
    tree = model.tree_
    if isinstance(model.classes_, list):
        # 多出力分類のときは classes_ がリストになる
        features = APP_DATASET_ONE_HOT[CATEGORICAL_FEATURES]
        labels = APP_DATASET_ONE_HOT[CATEGORICAL_LABELS]
    else:
        dataset = _numeric_dataset()
        features = dataset[NUMERIC_FEATURES]
        labels = dataset["App"]

    return TreeSummary(
        root_feature=feature_names[tree.feature[0]],
        root_threshold=float(tree.threshold[0]),
        root_gini=float(tree.impurity[0]),
        node_count=int(tree.node_count),
        score=float(model.score(features, labels)),
    )


def split_features(model: DecisionTreeClassifier, feature_names: list[str]) -> list[str]:
    """木のなかで実際に分割に使われた特徴量を、節の順に並べる。

    葉は `-2` で表されるので取り除く。同点の候補から無作為に選ばれるため、
    ここが実行のたびに変わる。
    """
    return [feature_names[index] for index in model.tree_.feature if index >= 0]

File: apps/grokking-ml-kotlin-lib/src/main/kotlin/lib/Nb09AppRecommendations.kt
package lib

import smile.base.cart.SplitRule
import smile.classification.DecisionTree
import smile.data.DataFrame
import smile.data.formula.Formula
import smile.data.type.DataTypes
import smile.data.type.StructField
import smile.data.vector.IntVector

/**
 * 原著ノートブック #09 `Chapter_09_Decision_Trees/App_recommendations.ipynb`。
 *
 * 6 人のユーザーの「使っている端末」と「年齢」から、おすすめアプリを決定木で当てる。
 * 原著は同じデータを 2 通りの形で学習させ、木の形がどう変わるかを見せている。
 *
 * 1. 年齢を **カテゴリ**（若者 / 大人）に潰して one-hot 符号化する
 * 2. 年齢を **数値** のまま渡す
 *
 * scikit-learn と Smile では決定木の作り方が違う点が 2 つある。
 * - Smile は **多出力分類に対応していない**。カテゴリ版は 1 列のラベルに直す
 * - Smile の `DecisionTree` は既定で決定的に分割する。scikit-learn のように
 *   同点の候補から無作為に選ぶことはない
 */
object Nb09AppRecommendations {

    /** 原著の元データ。この 6 人ぶんの情報しかない */
    val PLATFORMS = listOf("iPhone", "iPhone", "Android", "iPhone", "Android", "Android")
    val AGES = intArrayOf(15, 25, 32, 35, 12, 14)
    val APPS = listOf(
        "Atom Count", "Check Mate Mate", "Beehive Finder",
        "Check Mate Mate", "Atom Count", "Atom Count",
    )

    /** アプリ名を 0 始まりの番号に直したもの。Smile はラベルを Int で扱う */
    val APP_CLASSES = APPS.distinct().sorted()

    private fun appIndices(): IntArray = IntArray(APPS.size) { APP_CLASSES.indexOf(APPS[it]) }

    private fun buildFrame(
        columns: Map<String, IntArray>,
        labels: IntArray,
        classCount: Int,
    ): DataFrame {
        // Smile は目的変数が「名義尺度」であることを型で示す必要がある。
        // ただの IntVector にすると回帰木として扱われてしまう
        val labelField = StructField(
            "App",
            DataTypes.IntegerType,
            smile.data.measure.NominalScale(*Array(classCount) { APP_CLASSES[it] }),
        )
        val vectors = buildList {
            columns.forEach { (name, values) -> add(IntVector.of(name, values)) }
            add(IntVector.of(labelField, labels))
        }
        return DataFrame.of(*vectors.toTypedArray())
    }

    /**
     * 年齢をカテゴリに潰した版を学習する。
     *
     * 原著は 3 列の目的変数を同時に予測する **多出力分類** にしているが、
     * Smile は対応していない。3 つのアプリを 1 列の名義尺度に直して学習する。
     * 結果として木の形は原著と変わりうるが、全問正解する点は同じである。
     */
    fun fitCategorical(): DecisionTree {
        val frame = buildFrame(
            linkedMapOf(
                "Platform_iPhone" to intArrayOf(1, 1, 0, 1, 0, 0),
                "Platform_Android" to intArrayOf(0, 0, 1, 0, 1, 1),
                "Age_Adult" to intArrayOf(0, 1, 1, 1, 0, 0),
                "Age_Young" to intArrayOf(1, 0, 0, 0, 1, 1),
            ),
            appIndices(),
            APP_CLASSES.size,
        )
        return fitTree(frame)
    }

    /**
     * 年齢を数値のまま渡した版を学習する。
     *
     * 決定木がしきい値を自分で決めるので、`Age <= 20` のような分割が現れる。
     */
    fun fitNumeric(): DecisionTree {
        val frame = buildFrame(
            linkedMapOf(
                "Age" to AGES,
                "Platform_iPhone" to intArrayOf(1, 1, 0, 1, 0, 0),
                "Platform_Android" to intArrayOf(0, 0, 1, 0, 1, 1),
            ),
            appIndices(),
            APP_CLASSES.size,
        )
        return fitTree(frame)
    }

    /**
     * 決定木を学習する。
     *
     * 引数 2 つの `DecisionTree.fit` は、葉の数の上限をデータ件数から自動で決める。
     * 6 件しかないと上限が 1 になり `Invalid maximum leaves: 1` で落ちるので、
     * ここでは上限を明示的に渡す。scikit-learn は既定で制限なしに育てる。
     */
    private fun fitTree(frame: DataFrame): DecisionTree =
        DecisionTree.fit(Formula.lhs("App"), frame, SplitRule.GINI, 20, 100, 1)

    /** 学習データに対する正解率 */
    fun accuracy(model: DecisionTree, frame: DataFrame): Double {
        val labels = appIndices()
        val correct = (0 until frame.nrow()).count { model.predict(frame[it]) == labels[it] }
        return correct.toDouble() / labels.size
    }

    /**
     * 木の構造を DOT 形式の文字列で得る。
     *
     * scikit-learn は `tree.plot_tree` で図を描くが、Smile は
     * **Graphviz の DOT 形式を文字列で返す** だけである。図にするには
     * 別途 Graphviz に通す必要がある。テキストなので差分は取りやすい。
     */
    fun toDot(model: DecisionTree): String = model.dot()

    /**
     * DOT 文字列から、分割に使われた特徴量としきい値を根から順に取り出す。
     *
     * Smile のラベルは `<Age &le; 20.0<br/>size = 6<br/>...>` という HTML 風の
     * 書式になっている。`&le;` は `<=` の実体参照である。
     */
    fun splits(model: DecisionTree): List<Split> =
        SPLIT_LABEL.findAll(model.dot())
            .map { Split(it.groupValues[1], it.groupValues[2].toDouble()) }
            .toList()

    /** 分割に使われた特徴量としきい値 */
    data class Split(val feature: String, val threshold: Double)

    /** DOT のノードラベルから特徴量名としきい値を取り出す正規表現 */
    private val SPLIT_LABEL = Regex("""label=<([A-Za-z_]+) &le; ([0-9.]+)<br/>""")

    /** 葉が予測するクラス名を、DOT に現れる順に取り出す */
    fun leafClasses(model: DecisionTree): List<String> =
        LEAF_LABEL.findAll(model.dot()).map { it.groupValues[1] }.toList()

    private val LEAF_LABEL = Regex("""label=<App = ([^<]+)<br/>""")
}
